### PR TITLE
AMQP subscription

### LINKED ISSUE #4: Add class for subscribing to a queue
[Chapman](https://github.com/twingly/chapman/blob/master/lib/binlog/queue.rb), [Zambezi](https://github.com/twingly/zambezi/blob/master/lib/binlog/queue.rb), (and the [crawler](https://trello.com/c/7cCnJUbz)) etc. subscribes to queues. Would be good if the subscription could be moved into this gem :)

### PR DESCRIPTION
**Pull request by roback**
close #4

Added a class for subscribing to queues. The code is roughly the same as in chapman and zambezi.

Tested it by using `rabbitmqadmin`:

```shell
$ rabbitmqadmin declare exchange name=binlog-test type=topic

$ bundle console
> s = Twingly::AMQP::Subscription.new(queue_name: "testing-twingly-amqp", routing_key: "testing.#", exchange_topic: "binlog-test")
> s.subscribe do |message|
>   puts message
> end
```

```shell
$ rabbitmqadmin publish exchange=binlog-test routing_key="testing.test" payload='{"message": "hello!"}'
```

Tests for this would be nice, but I don't know how to do. I don't know how you should test `Subscription#subscribe` since it blocks the starting thread.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -19,13 +19,35 @@ gem "twingly-amqp", :git => "git@github.com:twingly/twingly-amqp.git"
...
 )
+```
+
+### Subscribe to a queue
+
+```ruby
+subscription = Twingly::AMQP::Subscription.new(
+  queue_name:       "queue-to-subscribe-to",
+  exchange_topic:   "exchange-name-to-connect-queue-to",
+  routing_key:      "blog.url.#",
```
Comment (walro) [inline README.md]: `queue_name` and `exchange_topic` is general in their example naming, `routing_key` is quite specific with `blog.url.#`. Perhaps use a "scenario" for all three or be general for all three

File: lib/twingly/amqp/subscription.rb [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+      def subscribe(&block)
+        queue    = @channel.queue(@queue_name, queue_options)
+        exchange = @channel.topic(@exchange_topic, durable: true)
+        queue.bind(exchange, routing_key: @routing_key)
+
+        setup_traps
+
+        consumer = queue.subscribe(subscribe_options) do |delivery_info, metadata, payload|
+          begin
+            Raven::Context.clear!
```
Comment (walro) [inline lib/twingly/amqp/subscription.rb]: Hmmm.. Raven or not that is the question =)

Comment (roback) [inline lib/twingly/amqp/subscription.rb]: Thought about that too, would be better not to have it as a dependency :).

Raven is used here too: https://github.com/twingly/twingly-amqp/pull/8/files#diff-dad5e21a79b90b462ebf9695f101a6dbR54 If we remove we have to expose methods in which the client can do this.

File: README.md (hunk 1 above)
Comment (roback) [inline README.md]: Yes, sure :)

File: lib/twingly/amqp/subscription.rb (hunk 2 above)
Comment (roback) [inline lib/twingly/amqp/subscription.rb]: Removed raven as dependency and added some callbacks instead: cff2acfade356a6f369e2e238700e9195aec4bea

**Comment by roback** [comment]
I'm finished with this one. I'll add tests for this and `#ping` in another branch instead of doing it here.

**Comment by walro** [comment]
Ah, forgot to say so, but it looked good to me :)

**Comment by jage** [comment]
:+1:

**Pull request merged**